pyro/infer/resampler.py

```diff
@@ -0,0 +1,134 @@
+    then one can reuse most of the previous samples via importance resampling.
+
+    :param callable guide: A pyro model that takes no arguments. The guide
+        should be diffuse, covering more space than the subsequent ``model``
+        passed to :meth:`sample`.
+    :param callable simulator: An optional larger pyro model with a superset of
+        the guide's latent variables.
+    :param int num_guide_samples: Number of inital samples to draw from the
+        guide. This should be much larger than the ``num_samples`` requested in
+        subsequent calls to :meth:`sample`.
+    :param int max_plate_nesting: The maximum plate nesting in the model.
+        If absent this will be guessed by running the guide.
+    """
+
+    def __init__(
+        self,
+        guide: Callable,
+        simulator: Optional[Callable] = None,
+        *,
+        num_guide_samples: int,
+        max_plate_nesting: Optional[int] = None,
+    ):
+        super().__init__()
+        if max_plate_nesting is None:
+            max_plate_nesting = _guess_max_plate_nesting(
+                guide if simulator is None else simulator
+            )
+        self._particle_dim = -1 - max_plate_nesting
+        self._gumbels: Optional[torch.Tensor] = None
+
+        # Draw samples from the initial guide.
+        with pyro.plate("particles", num_guide_samples, dim=self._particle_dim):
+            trace = poutine.trace(guide).get_trace()
+            self._old_logp = _log_prob_sum(trace, num_guide_samples)
+
+            if simulator:
+                # Draw extended samples from the simulator.
+                trace = poutine.trace(poutine.replay(simulator, trace)).get_trace()
+        self._samples = {
+            name: site["value"]
+            for name, site in trace.nodes.items()
+            if site["type"] == "sample" and not site_is_subsample(site)
+        }
+
+    @torch.no_grad()
+    def sample(
+        self, model: Callable, num_samples: int, stable: bool = True
+    ) -> Dict[str, torch.Tensor]:
+        """Draws a set of at most ``num_samples`` many model samples,
+        optionally extended by the ``simulator``.
+
+        Internally this importance resamples the samples generated by the
+        ``guide`` in ``.__init__()``, and does not rerun the ``guide`` or
+        ``simulator``. If the original guide samples poorly cover the model
+        distribution, samples will show low diversity.
+
+        :param callable model: A model with the same latent variables as the
+            original ``guide``.
+        :param int num_samples: The number of samples to draw.
+        :param bool stable: Whether to use piecewise-constant multinomial
+            sampling. Set to True for visualization, False for Monte Carlo
+            integration. Defaults to True.
+        :returns: A dictionary of stacked samples.
+        :rtype: Dict[str, torch.Tensor]
+        """
+        # Importance sample: keep all weights >= 1; subsample weights < 1.
```

Comment:
what does this comment mean?

Reply:
oops, this is obsolete. I've deleted it.